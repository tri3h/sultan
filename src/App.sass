@import "./styles/fonts.css"
@import "./styles/reset.css"
@import "./styles/variables.sass"

* 
    font-family: $font-inter
    font-style: normal