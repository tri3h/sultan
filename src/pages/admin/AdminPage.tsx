import { FC, useState } from "react";
import classes from "./AdminPage.module.sass";
import AdminProductCard from "../../components/adminProductCard/AdminProductCard";
import { LOCAL_STORAGE_DATA, Product } from "../../types/types";
import { getProductsFromStorage, getTypesFromStorage } from "../../utils/utils";

const AdminPage: FC = () => {
    const products = getProductsFromStorage();
    const types = getTypesFromStorage();
    const addingDefaultValues = {
        name: "",
        size: "",
        barcode: "",
        producer: "",
        brand: "",
        description: "",
        price: "",
        type: products[0].type,
        care_type: "",
        pic: products[0].pic,
    };
    const [editingDefaultValues, setEditingDefaultValues] =
        useState(addingDefaultValues);
    const [careTypeSelected, setCareTypeSelected] = useState("");
    const [careTypeChanged, setCareTypeChanged] = useState("");
    const productOptions = products.map((product) => (
        <option key={product.barcode} value={product.barcode}>
            {product.name}
        </option>
    ));
    const deleteProduct = () => {
        const barcode = editingDefaultValues.barcode;
        const newProducts = products.filter(
            (product) => product.barcode !== barcode
        );
        localStorage.setItem(
            LOCAL_STORAGE_DATA.PRODUCTS,
            JSON.stringify(newProducts)
        );
        setEditingDefaultValues(addingDefaultValues);
    };
    const onChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
        const barcode = event.currentTarget.value;
        const product = products.find((p) => p.barcode === barcode);

        if (product) {
            setEditingDefaultValues({
                name: product.name,
                size: product.size,
                barcode: product.barcode,
                producer: product.producer,
                brand: product.brand,
                description: product.description,
                price: product.price,
                type: product.type,
                care_type: "",
                pic: product.pic,
            });
        }
    };
    const careTypesOptions = types.map((type) => (
        <option key={type} value={type}>
            {type}
        </option>
    ));
    const onCareTypeChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
        setCareTypeSelected(event.currentTarget.value);
        setCareTypeChanged(event.currentTarget.value);
    };
    const onCareTypeInputChange = (
        event: React.ChangeEvent<HTMLInputElement>
    ) => {
        setCareTypeChanged(event.currentTarget.value);
    };
    const saveCareType = () => {
        let newTypes = types.filter((type) => type !== careTypeSelected);
        newTypes.push(careTypeChanged);
        let newProducts = products.map((product) => {
            let changedTypes = product.care_type.split(", ");
            changedTypes = changedTypes.filter(
                (type) => type !== careTypeSelected
            );
            changedTypes.push(careTypeChanged);
            let result = changedTypes.join(", ");
            return { ...product, care_type: result };
        });
        localStorage.setItem(
            LOCAL_STORAGE_DATA.CARE_TYPES,
            JSON.stringify(newTypes)
        );
        localStorage.setItem(
            LOCAL_STORAGE_DATA.PRODUCTS,
            JSON.stringify(newProducts)
        );
    };
    const addNewProduct = (product: Product) => {
        products.push(product);
        localStorage.setItem(
            LOCAL_STORAGE_DATA.PRODUCTS,
            JSON.stringify(products)
        );
    };
    const editProduct = (product: Product) => {
        let newProducts = products.filter((p) => p.barcode !== product.barcode);
        newProducts.push(product);
        localStorage.setItem(
            LOCAL_STORAGE_DATA.PRODUCTS,
            JSON.stringify(newProducts)
        );
    };
    return (
        <div className={classes.container}>
            <div>
                <h1 className={classes.title}>Добавление товара</h1>
                <AdminProductCard
                    defaultValues={addingDefaultValues}
                    save={addNewProduct}
                />
            </div>
            <div>
                <h1 className={classes.title}>Редактирование типов ухода</h1>
                <p className={classes.text}>
                    Выберите тип ухода для редактирования:
                </p>
                <div className={classes["care-type"]}>
                    <select
                        className={classes.select}
                        onChange={onCareTypeChange}
                    >
                        <option disabled selected hidden></option>
                        {careTypesOptions}
                    </select>
                    <input
                        className={classes.select}
                        value={careTypeChanged}
                        onChange={onCareTypeInputChange}
                    ></input>
                    <button className={classes.button} onClick={saveCareType}>
                        Сохранить
                    </button>
                </div>
            </div>
            <div>
                <h1 className={classes.title}>Редактирование товаров</h1>
                <div className={classes.selecting}>
                    <p className={classes.text}>
                        Выберите товар для редактирования:
                    </p>
                    <select
                        className={[
                            classes.select,
                            classes["product-editing-select"],
                        ].join(" ")}
                        onChange={onChange}
                    >
                        <option disabled selected hidden></option>
                        {productOptions}
                    </select>
                    <AdminProductCard
                        defaultValues={editingDefaultValues}
                        save={editProduct}
                    />
                    <button className={classes.button} onClick={deleteProduct}>
                        УДАЛИТЬ ВЫБРАННЫЙ ТОВАР
                    </button>
                </div>
            </div>
        </div>
    );
};

export default AdminPage;
