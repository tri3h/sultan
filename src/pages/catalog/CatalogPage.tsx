import { FC, useEffect } from "react";
import classes from "./CatalogPage.module.sass";
import Header from "../../components/header/Header";
import Footer from "../../components/footer/Footer";

import ProductList from "../../components/productList/ProductList";
import Sorting from "../../components/sorting/Sorting";
import FilterProducer from "../../components/filterProducer/filterProducer";
import FilterPrice from "../../components/filterPrice/filterPrice";
import FilterType from "../../components/filterType/filterType";
import ShadowBlock from "../../components/shadowBlock/ShadowBlock";
import Breadcrumbs from "../../components/breadcrumbs/Breadcrumbs";
import { useTypedSelector } from "../../hooks/useTypedSelector";
import { useActions } from "../../hooks/useActions";
import { Product } from "../../types/types";

const CatalogPage: FC = () => {
    const { care_types, products } = useTypedSelector((state) => state.product);
    const breadcrumbs = [
        { name: "Главная", url: "/" },
        { name: "Косметика и гигиена", url: "/" },
    ];
    const shadowBlocks = care_types.map((type, index) => (
        <li key={index}>
            <ShadowBlock name={type} />
        </li>
    ));
    const {
        producer_search,
        producer_checkboxes,
        care_type_checkboxes,
        price_start,
        price_end,
    } = useTypedSelector((state) => state.filter);
    const { setFilteredProducts, setProductPage } = useActions();
    const filter = () => {
        let filtered = [...products];
        //filter by price
        filtered = filtered.filter((product) => {
            let price = Number(product.price);
            return price >= price_start && price <= price_end;
        });
        //filter by producer
        if (producer_checkboxes.length > 0) {
            filtered = filtered.filter((product: Product) =>
                producer_checkboxes.includes(product.producer)
            );
        }
        if (producer_search.length > 0) {
            filtered = filtered.filter((product: Product) =>
                product.producer
                    .toLowerCase()
                    .includes(producer_search.toLowerCase())
            );
        }
        //filter by care type
        if (care_type_checkboxes.length > 0) {
            filtered = filtered.filter((product: Product) => {
                let care = product.care_type.split(", ");
                return care_type_checkboxes.some((checkbox) =>
                    care.includes(checkbox)
                );
            });
        }

        setFilteredProducts(filtered);
        setProductPage(1);
    };
    useEffect(() => {
        filter();
    }, [
        producer_search,
        producer_checkboxes,
        care_type_checkboxes,
        price_start,
        price_end,
    ]);
    return (
        <>
            <Header />
            <main className={classes.main}>
                <Breadcrumbs links={breadcrumbs} />
                <div className={classes.top}>
                    <h1 className={classes.title}>Косметика и гигиена</h1>
                    <Sorting />
                </div>
                <ul className={classes["shadow-blocks"]}>{shadowBlocks}</ul>
                <div className={classes.content}>
                    <div className={classes.filters}>
                        <h2 className={classes.subtitle}>
                            ПОДБОР ПО ПАРАМЕТРАМ
                        </h2>
                        <FilterPrice />
                        <FilterProducer />
                        <FilterType />
                    </div>
                    <div className={classes.data}>
                        <ProductList />
                        <div className={classes.text}>
                            Lorem ipsum dolor sit amet, consectetur adipiscing
                            elit. Nullam interdum ut justo, vestibulum sagittis
                            iaculis iaculis. Quis mattis vulputate feugiat massa
                            vestibulum duis. Faucibus consectetur aliquet sed
                            pellentesque consequat consectetur congue mauris
                            venenatis. Nunc elit, dignissim sed nulla
                            ullamcorper enim, malesuada.
                        </div>
                    </div>
                </div>
            </main>
            <Footer />
        </>
    );
};

export default CatalogPage;
