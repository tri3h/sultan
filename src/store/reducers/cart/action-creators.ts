import { CartAction, CartActionEnum, CartItem } from "./types";

export const CartActionCreators = {
    setCartItem: (items: CartItem[]): CartAction => ({
        type: CartActionEnum.SET_CART_ITEM,
        payload: items,
    }),
};
