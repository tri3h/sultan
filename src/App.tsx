import { FC } from "react";
import "./App.sass";
import AppRouter from "./components/appRouter/AppRouter";

const App: FC = () => {
    return <AppRouter />;
};

export default App;
