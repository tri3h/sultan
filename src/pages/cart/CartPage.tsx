import { FC, useState } from "react";
import classes from "./CartPage.module.sass";
import CartList from "../../components/cartList/CartList";
import Header from "../../components/header/Header";
import Footer from "../../components/footer/Footer";
import CustomButton from "../../components/button/CustomButton";
import Breadcrumbs from "../../components/breadcrumbs/Breadcrumbs";
import { useTypedSelector } from "../../hooks/useTypedSelector";
import ThanksModal from "../../components/thanksModal/ThanksModal";
import { useActions } from "../../hooks/useActions";
import { RouteNames } from "../../routes";

const CartPage: FC = () => {
    const { items } = useTypedSelector((state) => state.cart);
    const [isBought, setIsBought] = useState(false);
    const { setCartItem } = useActions();
    const totalPrice = items.reduce(
        (prev, item) => Number(item.count) * Number(item.product.price) + prev,
        0
    );
    const onBuyClick = () => {
        setCartItem([]);
        window.scrollTo(0, 0);
        setIsBought(true);
        document.body.style.overflow = "hidden";
    };
    const closeModal = () => {
        setIsBought(false);
        document.body.style.overflow = "";
    };
    const breadcrumbs = [
        { name: "Главная", url: RouteNames.CATALOG },
        { name: "Корзина", url: RouteNames.CART },
    ];
    const noItems = (
        <div className={classes["no-items"]}>В корзине нет товаров</div>
    );
    const itemsList = (
        <div className={classes.cart}>
            <CartList />
            <div className={classes.final}>
                <CustomButton text={"Оформить заказ"} onClick={onBuyClick} />
                <p className={classes.price}>{`${totalPrice} ₸`}</p>
            </div>
        </div>
    );
    return (
        <>
            <Header />
            <main data-testid="cart-page-main" className={classes.main}>
                <Breadcrumbs links={breadcrumbs} />
                <h1 className={classes.title}>Корзина</h1>
                {items.length === 0 ? noItems : itemsList}
                {isBought ? <ThanksModal close={closeModal} /> : null}
            </main>
            <Footer />
        </>
    );
};

export default CartPage;
