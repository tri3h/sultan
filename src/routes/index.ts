import React from "react";
import CatalogPage from "../pages/catalog/CatalogPage";
import ProductPage from "../pages/product/ProductPage";
import CartPage from "../pages/cart/CartPage";
import AdminPage from "../pages/admin/AdminPage";

export interface IRoute {
    path: string;
    element: React.ComponentType;
}

export enum RouteNames {
    CATALOG = "/",
    PRODUCT = "/product/:barcode",
    CART = "/cart",
    ADMIN = "/admin",
}

export const routes: IRoute[] = [
    { path: RouteNames.CATALOG, element: CatalogPage },
    { path: RouteNames.PRODUCT, element: ProductPage },
    { path: RouteNames.CART, element: CartPage },
    { path: RouteNames.ADMIN, element: AdminPage },
];
