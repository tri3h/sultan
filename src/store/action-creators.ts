import { CartActionCreators } from "./reducers/cart/action-creators";
import { FilterActionCreators } from "./reducers/filter/action-creators";
import { ProductActionCreators } from "./reducers/product/action-creators";

export const allActionCreators = {
    ...ProductActionCreators,
    ...CartActionCreators,
    ...FilterActionCreators,
};
