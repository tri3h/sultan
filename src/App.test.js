import { fireEvent, render, screen } from "@testing-library/react";
import App from "./App";
import CatalogPage from "./pages/catalog/CatalogPage";
import { Provider } from "react-redux";
import { store } from "./store";
import { MemoryRouter } from "react-router-dom";
import AdminPage from "./pages/admin/AdminPage";
import CartPage from "./pages/cart/CartPage";
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";
import { RouteNames } from "./routes";
import filterReducer from "./store/reducers/filter";
import * as utils from "./utils/utils";
import * as filterActions from "./store/reducers/filter/action-creators";

describe("tests", () => {
    beforeEach(() => {
        jest.spyOn(utils, "getImageByName").mockImplementation(() => "");
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    test("Header is present on the cart page", () => {
        render(
            <Provider store={store}>
                <MemoryRouter>
                    <CartPage />
                </MemoryRouter>
            </Provider>
        );
        const header = screen.queryByTestId("header");
        expect(header).toBeInTheDocument();
    });

    test("Expand btn hides after being clicked", () => {
        render(
            <Provider store={store}>
                <MemoryRouter>
                    <CatalogPage />
                </MemoryRouter>
            </Provider>
        );
        const btn = screen.getByTestId("checkbox-list-expand");
        fireEvent.click(btn);
        expect(btn).not.toBeInTheDocument();
    });

    test("Input value changes when user types", async () => {
        render(
            <Provider store={store}>
                <MemoryRouter>
                    <CatalogPage />
                </MemoryRouter>
            </Provider>
        );
        const search = screen.getAllByPlaceholderText("Поиск...")[1];
        await act(async () => {
            userEvent.type(search, "123");
        });
        expect(search.value).toBe("123");
    });

    test("Error shows when saving product with empty fields", () => {
        render(
            <Provider store={store}>
                <MemoryRouter>
                    <AdminPage />
                </MemoryRouter>
            </Provider>
        );
        const btn = screen.getAllByTestId("admin-save-btn")[0];
        fireEvent.click(btn);
        const error = screen.queryByText("Все поля должны быть заполнены");
        expect(error).toBeInTheDocument();
    });

    test("Cart matches snapshot", () => {
        render(
            <Provider store={store}>
                <MemoryRouter>
                    <CatalogPage />
                </MemoryRouter>
            </Provider>
        );
        const cart = screen.getByTestId("cart-icon-container");
        expect(cart).toMatchSnapshot();
    });

    test("Open catalog page when route is incorrect", () => {
        render(
            <Provider store={store}>
                <MemoryRouter initialEntries={["/123456"]}>
                    <App />
                </MemoryRouter>
            </Provider>
        );
        expect(screen.getByTestId("catalog-page-main")).toBeInTheDocument();
    });

    test("Click on the cart icon in header opens cart page", () => {
        render(
            <Provider store={store}>
                <MemoryRouter initialEntries={[RouteNames.CATALOG]}>
                    <App />
                </MemoryRouter>
            </Provider>
        );
        const cartIcon = screen.getByTestId("cart-icon-container");
        fireEvent.click(cartIcon);
        expect(screen.queryByTestId("catalog-page-main")).toBeNull();
        expect(screen.getByTestId("cart-page-main")).toBeInTheDocument();
    });

    test("Succesfully change product page in filter reducer", () => {
        const state = filterReducer(
            undefined,
            filterActions.FilterActionCreators.setProductPage(2)
        );
        expect(state.page).toBe(2);
    });

    test("Click on care type block changes care type checkbox", () => {
        render(
            <Provider store={store}>
                <MemoryRouter>
                    <CatalogPage />
                </MemoryRouter>
            </Provider>
        );
        const checkbox = screen.getAllByTestId("checkbox-input")[0];
        const block = screen.getAllByTestId("block-item")[0];
        expect(checkbox).not.toBeChecked();
        fireEvent.click(block);
        expect(checkbox).toBeChecked();
    });

    test("Chosen option in select shows as default value in input", async () => {
        render(
            <Provider store={store}>
                <MemoryRouter>
                    <AdminPage />
                </MemoryRouter>
            </Provider>
        );
        const select = screen.getByTestId("admin-type-select");
        const option = screen.getAllByTestId("admin-type-option")[1];
        await act(async () => {
            userEvent.selectOptions(select, option.value);
        });
        const input = screen.getByTestId("admin-type-input");
        expect(option.value).toBe(input.value);
    });
});
