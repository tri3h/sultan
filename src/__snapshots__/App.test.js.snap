// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests Cart matches snapshot 1`] = `
<div
  class="container"
  data-testid="cart-icon-container"
>
  <div
    class="pic-container"
  >
    <span
      class="counter"
    >
      0
    </span>
    <img
      alt="cart"
      src="cart.svg"
    />
  </div>
  <div
    class="text-container"
  >
    <span
      class="text"
    >
      Корзина
    </span>
    <span
      class="price"
    >
      0 ₸
    </span>
  </div>
</div>
`;
